src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -9223,6 +9223,10 @@ private ExpressionSyntax ParseTerm(Precedence precedence)
                 case SyntaxKind.DelegateKeyword:
                     expr = this.ParseAnonymousMethodExpression();
                     break;
+                case SyntaxKind.RefKeyword:
+                    // ref is not expected to appear in this position.
+                    expr = this.AddError(ParsePossibleRefExpression(), ErrorCode.ERR_InvalidExprTerm, SyntaxFacts.GetText(tk));
```

Comment:
it worries me that we still hve infinite loop holes here.  It seems like we're getting into parseterm, but there's no guarantee that parseterm will make progress.  and the caller loops without checking for that.  either ParseTerm should always make progrss (probably a bad idea), or the caller must ensure that if it's not making progress it skips a token/tokens and continues on.

Reply:
It looks like the parser has an `IsMakingProgress` method which is meant to catch this. I will take a look at why it did not catch it in this scenario.

Reply:
Opened #39331 as part of this work